### PR TITLE
HDDS-4983. Display key offset for each block in command key info

### PR DESCRIPTION
**Pull request by cchenax**
## What changes were proposed in this pull request?

add keyOffset in the key info result

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4983

## How was this patch tested?
example:sh key info /vol1/bucket1/key3

```
{
  "volumeName" : "vol1",
  "bucketName" : "bucket1",
  "name" : "key3",
  "dataSize" : 644930593,
  "creationTime" : "2021-04-20T09:22:15.206Z",
  "modificationTime" : "2021-04-20T09:22:20.869Z",
  "replicationType" : "RATIS",
  "replicationFactor" : 3,
  "ozoneKeyLocations" : [ {
    "containerID" : 3,
    "localID" : 107544261427200003,
    "length" : 268435456,
    "offset" : 0,
    "keyOffset" : 0
  }, {
    "containerID" : 4,
    "localID" : 107544261427200004,
    "length" : 268435456,
    "offset" : 0,
    "keyOffset" : 268435456
  }, {
    "containerID" : 5,
    "localID" : 107544261427200005,
    "length" : 108059681,
    "offset" : 0,
    "keyOffset" : 536870912
  } ],
  "metadata" : { },
  "fileEncryptionInfo" : null
}
```

### CODE REVIEW

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneKeyLocation.java [hunk 1]
```diff
@@ -39,16 +39,20 @@
    * Offset of this key.
    */
   private final long offset;
-
+  /**
+   * KeyOffset of this key.
+   */
+  private final long KeyOffset;
```
Comment (cku328) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneKeyLocation.java:45]:
```suggestion
  private final long keyOffset;
```
use lower camel case: KeyOffset -> keyOffset

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneKeyLocation.java [hunk 2]
```diff
@@ -39,16 +39,20 @@
...
   /**
    * Constructs OzoneKeyLocation.
    */
   public OzoneKeyLocation(long containerID, long localID,
-                  long length, long offset) {
+                          long length, long offset) {
```
Comment (cku328) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneKeyLocation.java:50]: Nit: seems like an unnecessary adjustment.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneKeyLocation.java [hunk 3]
```diff
@@ -79,4 +83,9 @@ public long getOffset() {
     return offset;
   }
 
+  /**
+   * Returns the KeyOffset of this Key.
+   */
+  public long getKeyOffset() { return KeyOffset; }
```
Comment (cku328) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneKeyLocation.java:89]:
```suggestion
  public long getKeyOffset() { 
    return keyOffset;
  }
```
blocks should have line breaks.

**Review by cku328 — COMMENTED** [review]
Thanks @cchenax for working on this.
I have some suggestions for `checkstyle` not to pass.

**Comment by elek** [comment]
For the records: jira issue has a conversation about improving the initial implementation:

image

**Comment by cchenax** [comment]
> For the records: jira issue has a conversation about improving the initial implementation:
> 
> image

yes,I will improve it.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 4]
```diff
@@ -858,9 +859,17 @@ public OzoneKeyDetails getKeyDetails(
...
+    long lastLength = 0L;
+    List<OmKeyLocationInfo> omKeyLocationInfos = keyInfo
+        .getLatestVersionLocations().getBlocksLatestVersionOnly();
+    for (OmKeyLocationInfo info: omKeyLocationInfos) {
+      ozoneKeyLocations.add(new OzoneKeyLocation(info.getContainerID(),
+          info.getLocalID(), info.getLength(), info.getOffset(),
+          lastKeyOffset + lastLength));
+      lastKeyOffset = lastLength + lastKeyOffset;
+      lastLength = info.getLength();
+    }
```
Comment (elek) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:872]: Fix me if I am wrong, but based on my understanding the variable usage can be simplified in this way:

```suggestion
    long lastKeyOffset = 0L;
    List<OmKeyLocationInfo> omKeyLocationInfos = keyInfo
        .getLatestVersionLocations().getBlocksLatestVersionOnly();
    for (OmKeyLocationInfo info: omKeyLocationInfos) {
      ozoneKeyLocations.add(new OzoneKeyLocation(info.getContainerID(),
          info.getLocalID(), info.getLength(), info.getOffset(),
          lastKeyOffset));
      lastKeyOffset += info.getLength();
    }
```

**Review by elek — APPROVED** [review]
Thanks the patch @cchenax 

Overall it looks good to me, I checked the class and it's used only on the client side, safe to modify.

I have one question about simplifying the loop.

**Comment by cchenax** [comment]
> Thanks the patch @cchenax
> 
> Overall it looks good to me, I checked the class and it's used only on the client side, safe to modify.
> 
> I have one question about simplifying the loop.

ok,this way is good,thank you very much.

**Review by elek — APPROVED** [review]
+1, thanks the update @cchenax

**Pull request merged**